File: mobile/src/main/java/com/siliconlabs/bledemo/Browser/Adapters/SavedSearchesAdapter.java
package com.siliconlabs.bledemo.Browser.Adapters;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.Browser.Model.SavedSearch;
import com.siliconlabs.bledemo.Utils.FilterDeviceParams;
import com.siliconlabs.bledemo.Utils.SharedPrefUtils;

import java.util.HashMap;
import java.util.List;

public class SavedSearchesAdapter extends RecyclerView.Adapter<SavedSearchesAdapter.SavedSearchesViewHolder> {
    private List<SavedSearch> savedSearchList;
    private Context context;
    private SavedSearchesCallback savedSearchesCallback;
    private SharedPrefUtils sharedPrefUtils;

    public interface SavedSearchesCallback {
        void onClick(String name);
    }

    public class SavedSearchesViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView searchedTextTV;
        LinearLayout savedSearch;
        ImageButton removeBtn;

        public SavedSearchesViewHolder(View itemView) {
            super(itemView);
            searchedTextTV = itemView.findViewById(R.id.textview_saved_search);
            savedSearch = itemView.findViewById(R.id.saved_search);
            removeBtn = itemView.findViewById(R.id.removeBtn);
            savedSearch.setOnClickListener(this);
            removeBtn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    removeAt(getAdapterPosition());
                }
            });
        }

        @Override
        public void onClick(View v) {
            savedSearchesCallback.onClick(String.valueOf(searchedTextTV.getText()));
        }

    }

    public void removeAt(int position) {
        if (position < 0) return;
        HashMap<String, FilterDeviceParams> map = sharedPrefUtils.getMapFilter();
        map.remove(savedSearchList.get(position).getSearchText());
        sharedPrefUtils.updateMapFilter(map);
        savedSearchList.remove(position);
        notifyItemRemoved(position);
        notifyItemRangeChanged(position, savedSearchList.size());
    }

    public SavedSearchesAdapter(List savedSearchList, Context context, SavedSearchesCallback savedSearchesCallback) {
        this.savedSearchList = savedSearchList;
        this.context = context;
        this.savedSearchesCallback = savedSearchesCallback;
        this.sharedPrefUtils = new SharedPrefUtils(context);
    }

    @Override
    public void onBindViewHolder(SavedSearchesAdapter.SavedSearchesViewHolder holder, int position) {
        SavedSearch savedSearch = savedSearchList.get(position);
        holder.searchedTextTV.setText(savedSearch.getSearchText());

        FilterDeviceParams lastFilterDeviceParams = sharedPrefUtils.getLastFilter();
        if (lastFilterDeviceParams != null && lastFilterDeviceParams.getFilterName().toLowerCase().equals(savedSearch.getSearchText().toLowerCase())) {
            holder.searchedTextTV.setTextColor(ContextCompat.getColor(context, R.color.silabs_blue_selected));
        } else {
            holder.searchedTextTV.setTextColor(ContextCompat.getColor(context, R.color.silabs_subtle_text));
        }
    }

    @Override
    public SavedSearchesAdapter.SavedSearchesViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_saved_searches, parent, false);
        return new SavedSearchesViewHolder(view);
    }


    @Override
    public int getItemCount() {
        return savedSearchList.size();
    }

    public void addItem(SavedSearch savedSearch) {
        savedSearchList.add(0, savedSearch);
        notifyDataSetChanged();
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Browser/Adapters/SavedSearchesAdapter.kt
package com.siliconlabs.bledemo.Browser.Adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.siliconlabs.bledemo.Browser.Adapters.SavedSearchesAdapter.SavedSearchesViewHolder
import com.siliconlabs.bledemo.Browser.Models.SavedSearch
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.Utils.FilterDeviceParams
import com.siliconlabs.bledemo.Utils.SharedPrefUtils
import java.util.*

class SavedSearchesAdapter(private val savedSearchList: ArrayList<SavedSearch>, private val context: Context, private val savedSearchesCallback: SavedSearchesCallback) : RecyclerView.Adapter<SavedSearchesViewHolder>() {
    private val sharedPrefUtils = SharedPrefUtils(context)

    interface SavedSearchesCallback {
        fun onClick(name: String?)
    }

    inner class SavedSearchesViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private var tvSearchedText = itemView.findViewById(R.id.textview_saved_search) as TextView
        private var llSavedSearch = itemView.findViewById(R.id.saved_search) as LinearLayout
        private var btnRemove = itemView.findViewById(R.id.removeBtn) as ImageButton

        override fun onClick(v: View) {
            savedSearchesCallback.onClick(tvSearchedText.text.toString())
        }

        fun bind(savedSearch: SavedSearch) {
            tvSearchedText.text = savedSearch.searchText
            val lastFilterDeviceParams = sharedPrefUtils.lastFilter
            if (lastFilterDeviceParams != null && lastFilterDeviceParams.filterName.toLowerCase(Locale.getDefault()) == savedSearch.searchText.toLowerCase(Locale.getDefault())) {
                tvSearchedText.setTextColor(ContextCompat.getColor(context, R.color.silabs_blue_selected))
            } else {
                tvSearchedText.setTextColor(ContextCompat.getColor(context, R.color.silabs_subtle_text))
            }

            llSavedSearch.setOnClickListener(this)
            btnRemove.setOnClickListener { removeAt(adapterPosition) }
        }
    }

    fun removeAt(position: Int) {
        if (position < 0) return
        val map: HashMap<String, FilterDeviceParams?> = sharedPrefUtils.mapFilter
        map.remove(savedSearchList[position].searchText)
        sharedPrefUtils.updateMapFilter(map)
        savedSearchList.removeAt(position)
        notifyItemRemoved(position)
        notifyItemRangeChanged(position, savedSearchList.size)
    }

    override fun onBindViewHolder(holder: SavedSearchesViewHolder, position: Int) {
        val savedSearch = savedSearchList[position]
        holder.bind(savedSearch)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SavedSearchesViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_saved_searches, parent, false)
        return SavedSearchesViewHolder(view)
    }

    override fun getItemCount(): Int {
        return savedSearchList.size
    }

    fun addItem(savedSearch: SavedSearch) {
        savedSearchList.add(0, savedSearch)
        notifyDataSetChanged()
    }

}